### PR TITLE
💄[/view/_iscnId] Modify Layout of ISCN information

### PR DESCRIPTION
**Pull request by AuroraHuang22**
**view/_iscnId**
- [x] Pass transaction info to `view/iscnId` [task](https://www.notion.so/likecoin/transaction-on-Big-Dipper-b025361ec9ac49eca557d8f65e89407e) 
- [x] The way to display raw data [task](https://www.notion.so/likecoin/View-raw-data-6703df128ca94398aca4ad893d5ad99c)
- [x] Update author/creator/stakeholder related UI [task](https://www.notion.so/likecoin/Update-author-creator-stakeholder-related-UI-7bc8bb7f60d44ea09a38d1492e898093)
- [x] Click to copy wallet address [task](https://www.notion.so/likecoin/Click-to-copy-the-address-3c9f5f9545c943de80a1bb6f8fca47ca)
截圖 2021-09-16 上午10 41 26

**IscnRegisterForm**
- [x] likerID [task](https://www.notion.so/likecoin/Update-Liker-ID-input-6c895270035e4c669a655a41d993539d)
- [x] description & multiple url & wallet input fields [task](https://www.notion.so/likecoin/Add-description-field-and-others-c178de0e3dd54fda9d964321e3259bf1)
- [x] likerID dialog -> 預計等做 autofill 功能再一起做 [task](https://www.notion.so/likecoin/LikerID-Dialog-4855a0b039534a39981002305476a952)
>- [x] Name selector [task](https://www.notion.so/likecoin/Name-Selector-options-7f74a05d4b664fe2947b37922d6b46e8)
>- [x] Name selector 的 options 資料 -> 目前留空 [task](https://www.notion.so/likecoin/Name-Selector-options-7f74a05d4b664fe2947b37922d6b46e8)
截圖 2021-09-16 上午10 12 55
截圖 2021-09-16 上午10 42 52

### CODE REVIEW

File: components/Snackbar.vue [hunk 1]
```diff
@@ -51,6 +51,9 @@ export default class Snackbar extends Vue {
   // Preset of the Snackbar
   @Prop(String) readonly preset!: string
 
+  // Set timeout to close the Snackbar
+  @Prop(Number) readonly timeout!: number
```
Comment (nwingt) [inline components/Snackbar.vue:55]: Can provide default is 0

File: components/Snackbar.vue [hunk 2]
```diff
@@ -76,6 +79,13 @@ export default class Snackbar extends Vue {
   @ModelSync('open', 'toggle', { type: Boolean, default: false })
   isOpen!: boolean
 
+  @Watch('isOpen')
+  timeOut() {
+    if(this.timeout){
+      setTimeout(()=>{this.close()},this.timeout)
+    }
```
Comment (nwingt) [inline components/Snackbar.vue:86]: Please reformat

File: components/IscnUploadedInfo.vue [hunk 3]
```diff
@@ -93,6 +93,15 @@
       >
         <ContentFingerprintLink :item="ipfs" />
       </FormField>
+      <IconDiverMini class="mb-[24px]" />
+      <FormField
+        :label="$t('iscn.meta.transaction')"
+        class="mb-[12px]"
+      >
+        <Link :href="`https://node.iscn-dev-2.like.co/txs/${iscnHash}`">
```
Comment (nwingt) [inline components/IscnUploadedInfo.vue:101]: Should use different path on production, can place it on `constant/index.js`

**Review by nwingt — CHANGES_REQUESTED** [review]

File: constant/index.ts [hunk 4]
```diff
@@ -23,3 +23,10 @@ export const ISCN_GAS_FEE = 200000;
 export const ISCN_REGISTRY_NAME = 'likecoin-chain';
 
 export const ISCN_PREFIX = `iscn://${ISCN_REGISTRY_NAME}`;
+
+export const LIKER_LAND_URL = 'https://liker.land/'
+
+export const RAWDATA_URL_PRODUCTION = 'https://likecoin.nnkken.dev/txs?iscn_record.iscn_id='
```
Comment (williamchong) [inline constant/index.ts:29]: Use `https://mainnet-node.like.co/`

**Review by williamchong — CHANGES_REQUESTED** [review]

File: pages/view/_iscnId.vue [hunk 5]
```diff
@@ -227,5 +339,75 @@ export default class ViewIscnIdPage extends Vue {
...
+
+  get rawDataURL(){
+    return this.url + this.iscnId
+  }
+
+  async fetchTxHash() {
+    let txHash = await fetch(RAWDATA_URL_PRODUCTION + this.iscnId)
+      .then((res) => res.json())
+    if (!txHash.txs) {
+      txHash = await fetch(RAWDATA_URL_STAGING + this.iscnId)
```
[the remark this answers is not in the document]
Comment (williamchong) [reply pages/view/_iscnId.vue:390]: In fact the `RAWDATA_URL` constant can be defined as `const ISCN_TX_RAW_DATA_ENDPOINT  =  IS_TESTNET ? 'test.url' : 'prod.url'`

File: pages/view/_iscnId.vue [hunk 6]
```diff
@@ -227,5 +339,75 @@ export default class ViewIscnIdPage extends Vue {
...
+        .then((res) => res.json())
+        .then((res) => res.txs[0].txhash)
```
Comment (williamchong) [inline pages/view/_iscnId.vue:392]: Should filter by `{"key":"action","value":"create_iscn_record"}` since we could have other tx related to this iscn_id, should not hardcode `txs[0]`

**Review by williamchong — CHANGES_REQUESTED** [review]

File: utils/cosmos/iscn/sign.ts [hunk 7]
```diff
@@ -37,14 +40,30 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
             entity: {
-              '@id': authorId ? `did:cosmos:${authorId.slice(6)}` : authorUrl || authorName,
-              name: authorName,
-              url: authorUrl,
+              '@id': authorId || authorUrl || authorName,
+                name: authorName,
+                url: authorUrl,
+                authorWalletAddress,
+                likerId,
+                authorDescription,
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:66]: These keys should have same indentation as `@id`?

File: components/IscnUploadedInfo.vue [hunk 8]
```diff
@@ -146,6 +159,10 @@ export default class IscnUploadedInfo extends Vue {
     return `ipfs://${this.ipfsHash}`
   }
 
+   get transactionsURL(){
+     return BIG_DIPPER_TRANSACTIONS + this.iscnHash
```
Comment (williamchong) [inline components/IscnUploadedInfo.vue:163]: Prefer template string
```suggestion
     return `BIG_DIPPER_TRANSACTIONS${this.iscnHash}`
```

File: utils/cosmos/iscn/sign.ts [hunk 9]
```diff
@@ -37,14 +40,30 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+      authorWallets[i].forEach((a: any) => {
+        if(a.type === 'cosmos'){
+          authorId =`did:cosmos:${a.address.slice(6)}`
+        }else if (a.type ==='eth' && !authorId ){
+          authorId = `did:eth:${a.address}`
+        }else{
+          return ''
+        }
+        return undefined
+      })
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:53]: forEach does not need a return value
A simpler way to write this with forEach
```suggestion
      authorWallets[i].forEach((a: any) => {
        if (a.type === 'cosmos') {
          authorId =`did:cosmos:${a.address.slice(6)}`
        } else if (a.type ==='eth') {
          authorId = authorId || `did:eth:${a.address}`
        }
      })
```
Using reduce() would also work

File: pages/view/_iscnId.vue [hunk 10]
```diff
@@ -227,5 +350,86 @@ export default class ViewIscnIdPage extends Vue {
...
+  showStakeholder(index: number) {
+    this.isOpenAuthorDialog = true
+    this.stakeholderInfo.id = this.stakeholders[index].entity.id
+    this.stakeholderInfo.likerId = this.stakeholders[index].entity.likerId
+    this.stakeholderInfo.authorDescription =
+      this.stakeholders[index].entity.authorDescription
+    this.stakeholderInfo.authorWalletAddress =
+      this.stakeholders[index].entity.authorWalletAddress
+    this.stakeholderInfo.authorUrl = this.stakeholders[index].entity.url
+    this.stakeholderInfo.authorName = this.stakeholders[index].entity.name
```
Comment (williamchong) [inline pages/view/_iscnId.vue:432]:
```suggestion
    const {
      id,
      likerId,
      authorDescription,
      authorWalletAddress,
      url: authorUrl,
      name: authorName,
    } = this.stakeholders[index].entity;
    this.stakeholderInfo = {
      id,
      likerId,
      authorDescription,
      authorWalletAddress,
      authorUrl,
      authorName,
    }
```

File: utils/cosmos/iscn/sign.ts [hunk 11]
```diff
@@ -37,14 +40,30 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
       if (isNonEmpty) {
         stakeholders.push({
             entity: {
-              '@id': authorId ? `did:cosmos:${authorId.slice(6)}` : authorUrl || authorName,
-              name: authorName,
-              url: authorUrl,
+              '@id': authorId || authorUrl || authorName,
+                name: authorName,
+                url: authorUrl,
+                authorWalletAddress,
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:64]: Should split the address into `did:{type}:{address}`, and put into a `sameAs` array

File: utils/cosmos/iscn/sign.ts (hunk 7 above)
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:66]: should just be `description`

File: pages/view/_iscnId.vue [hunk 12]
```diff
@@ -227,5 +350,86 @@ export default class ViewIscnIdPage extends Vue {
...
+              attributes.value === 'create_iscn_record'
+            ) {
+              txhash = item[i].txhash
+            }
+          })
+        })
+      })
+    })
+    return txhash
+  }
```
Comment (williamchong) [inline pages/view/_iscnId.vue:421]: Should use a simple `for loop` + `.find` for early exit, `forEach` forces you to loop through everything

File: utils/cosmos/iscn/sign.ts [hunk 13]
```diff
@@ -37,14 +40,30 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+                likerId,
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:65]: LikerId should be written as a `url` to `https://like.co/{likerId}` instead of a field called `likerId`

**Review by williamchong — CHANGES_REQUESTED** [review]
Please do not rebase after someone has reviewed a pull request, would be very hard to follow what has changed sine last review

File: components/IscnRegisterForm.vue [hunk 14]
```diff
@@ -162,35 +169,48 @@
               size="mini"
               preset="secondary"
               content-class="py-[4px]"
-              @click="isOpenAuthorDialog = true"
+              @click="
+                () => {
+                  isOpenAuthorDialog = true
+                  initAuthorInfo()
+                }"
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:176]: Consider refactor to class method?

File: components/Snackbar.vue [hunk 15]
```diff
@@ -76,16 +79,22 @@ export default class Snackbar extends Vue {
...
   isOpen!: boolean
 
+  @Watch('isOpen')
+  timeOut() {
+    if (this.timeout) {
+      setTimeout(() => {
+        this.close()
+      }, this.timeout)
+    }
+  }
```
Comment (williamchong) [inline components/Snackbar.vue:89]: Do we need to clear this timeout if the user explicitly closed this? Would it cause error if `this.close()` is called when the snackbar is closed manually first?

File: components/Snackbar.vue (hunk 2 above)
Comment (nwingt) [inline components/Snackbar.vue]: Need to add life cycle method `beforeDestroy()` to remove the `clearTimeout()`

File: utils/cosmos/iscn/sign.ts [hunk 16]
```diff
@@ -25,26 +25,48 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+      authorWallets[i].forEach((a: any) => {
+        if (a.type === 'cosmos') {
+          authorId = `did:cosmos:${a.address.slice(6)}`
+          authorWalletAddress.push({ address: authorId, type: a.type })
+          sameAs.push(authorId)
+        } else if (a.type === 'eth') {
+          authorId = `did:eth:${a.address}`
+          authorWalletAddress.push({ address: authorId, type: a.type })
+          sameAs.push(authorId)
+        }
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:55]: Is there any `type` other than `cosmos` and `eth`? I think every wallets should be pushed into `sameAs` as `did:${type}:${address}`

File: constant/index.ts [hunk 17]
```diff
@@ -23,3 +23,10 @@ export const ISCN_GAS_FEE = 200000;
...
+export const RAWDATA_URL_STAGING = 'https://node.iscn-dev-2.like.co/txs?iscn_record.iscn_id='
+
+export const BIG_DIPPER_TRANSACTIONS = 'https://likecoin.bigdipper.live/transactions/'
```
Comment (nwingt) [inline constant/index.ts:35]:
```suggestion
export const BIG_DIPPER_TX_BASE_URL = 'https://likecoin.bigdipper.live/transactions/'
```

File: utils/cosmos/iscn/sign.ts [hunk 18]
```diff
@@ -25,26 +25,48 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+      })
+      const likerId = `https://like.co/${likerIds[i]}`
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:57]: This should either be used as `url` or `sameAs` instead of a new key called `likerId`
The `url` would means the primary url that can identify the author, where `sameAs` are all the other urls/ids that connects to the same author

File: utils/cosmos/iscn/sign.ts [hunk 19]
```diff
@@ -25,26 +25,48 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
       const isNonEmpty = authorUrl || authorName || authorId
       if (isNonEmpty) {
         stakeholders.push({
             entity: {
-              '@id': authorId ? `did:cosmos:${authorId.slice(6)}` : authorUrl || authorName,
+              '@id': authorId || authorUrl || authorName,
               name: authorName,
               url: authorUrl,
+              walletAddress: authorWalletAddress,
+              likerId,
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:67]: Should remove `walletAddress` and `likerId`, and use the content for `@id`, `url`, `sameAs` instead

File: components/TextField.vue [hunk 20]
```diff
@@ -37,10 +40,15 @@ export default class TextField extends Vue {
...
+  get tag(): any {
+    if (this.isTextarea) return 'textarea'
+    return 'input'
+  }
+
   get rootClasses(): any {
     return [
       TextField.presetClasses,
-      ...this.sizeClasses,
+      this.isTextarea? 'h-[62px] font-semibold resize-none': this.sizeClasses,
```
Comment (nwingt) [inline components/TextField.vue:51]:
```suggestion
      this.isTextarea ? 'h-[62px] font-semibold resize-none' : this.sizeClasses,
```

File: components/TextFieldList.vue [hunk 21]
```diff
@@ -47,25 +39,60 @@
...
-  fields: any = [{ key: 0 }]
+  @Model('change', { type: Array, default: () => [] }) value!: Array<object>
+
+  created() {
+    this.inputValue = this.value
+    if(!this.value.length){
+      this.value.push({
+      content: '',
+    })
+    }
```
Comment (nwingt) [inline components/TextFieldList.vue]: Please reformat

File: pages/view/_iscnId.vue [hunk 22]
```diff
@@ -189,29 +189,35 @@
...
+              <Link v-for="url in stakeholderInfo.authorUrl" :key="url.key" :href="url.content">{{ url.content }}</Link>
             </FormField>
             <IconDiverMini class="my-[12px]" />
             <!-- wallet address -->
-            <FormField :label="$t('iscn.meta.stakeholders.wallet')">
+            <FormField
+              v-if="stakeholderInfo.authorWalletAddress.length"
+              :label="$t('iscn.meta.stakeholders.wallet')"
+              content-classes="flex flex-row"
+              >
```
Comment (nwingt) [inline pages/view/_iscnId.vue:221]:
```suggestion
            >
```

File: pages/view/_iscnId.vue [hunk 23]
```diff
@@ -224,7 +230,7 @@
                 <template #prepend>{{
                   $t('iscn.meta.stakeholders.wallet.LIKE')
                 }}</template>
-                {{ stakeholderInfo.address.slice(4) | ellipsis }}
+                {{ address.content | ellipsis }}
```
Comment (nwingt) [inline pages/view/_iscnId.vue]:
```suggestion
                {{ address.content || ellipsis }}
```

**Review by nwingt — CHANGES_REQUESTED** [review]

File: pages/view/_iscnId.vue [hunk 24]
```diff
@@ -191,41 +390,80 @@ export default class ViewIscnIdPage extends Vue {
...
-        this.$nuxt.error({
-          statusCode: 400,
-          message: ErrorMessage.statusCode400,
-        })
-        return
-      }
-      this.iscnId = res.records[0].id
-      this.$router.replace({ params: { iscnId: this.iscnId } })
+  get rawDataURL() {
+    return `${ISCN_TX_RAW_DATA_ENDPOINT}${this.iscnId}`
```
Comment (williamchong) [inline pages/view/_iscnId.vue:398]: Should use ISCN query by ID endpoint `/iscn/records/id?iscn_id=${iscnId}` if you do not want the transaction hash, only the ISCN record data

File: pages/view/_iscnId.vue [hunk 25]
```diff
@@ -191,41 +390,80 @@ export default class ViewIscnIdPage extends Vue {
...
+  }
+
+  async fetchTxHash() {
+    const txHash = await this.$axios.get(`${ISCN_TX_RAW_DATA_ENDPOINT}${this.iscnId}`)
+      .then((res) =>this.getTxHash(res.data.txs))
```
Comment (williamchong) [inline pages/view/_iscnId.vue:403]: Why need `.then()` here when we are using `await`? 
```

    const res = await this.$axios.get(`${ISCN_TX_RAW_DATA_ENDPOINT}${this.iscnId}`)
    const txHash = this.getTxHash(res.data.txs)
```
Just await would do

File: pages/view/_iscnId.vue [hunk 26]
```diff
@@ -191,41 +390,80 @@ export default class ViewIscnIdPage extends Vue {
...
+          log.events.find((event: any) =>
+            event.attributes.some(
+              (attribute: any) =>
+                attribute.key === 'action' &&
+                attribute.value === 'create_iscn_record',
+            ),
+          ),
+        ),
+      )
+      return items[index].txhash
```
Comment (williamchong) [inline pages/view/_iscnId.vue:420]: Very clean code, but does `targetItem = items.find()` then `targetItem.txhash` also work, or does it have to be `findIndex()`?

File: pages/view/_iscnId.vue [hunk 27]
```diff
@@ -191,41 +390,80 @@ export default class ViewIscnIdPage extends Vue {
...
-      this.$nuxt.error({ statusCode: 404, message: ErrorMessage.statusCode404 })
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  handleCopy(address: string, type: string) {
+    let text = ''
+    if (type === 'cosmos') {
+      text = address.replace(/(did:|:)/g, '')
+    } else {
+      text = address.replace(/(did:|eth:|:)/g, '')
```
Comment (williamchong) [inline pages/view/_iscnId.vue:449]: If `cosmos` address is the only special case, I think we can just remove anything `did:*:` instead of hardcoding eth

**Review by williamchong — CHANGES_REQUESTED** [review]

File: pages/view/_iscnId.vue (hunk 23 above)
Comment (AuroraHuang22) [reply pages/view/_iscnId.vue]: `{{ address.content | ellipsis }}` 是 [Filters](https://vuejs.org/v2/guide/filters.html) 的用法，不是 or 的意思

File: components/EditableTagList.vue [hunk 28]
```diff
@@ -118,14 +124,17 @@ export default class EditableTagList extends Vue {
   }
 
   blurInput() {
+    if(this.currentInput === ' ') return
```
Comment (williamchong) [inline components/EditableTagList.vue:127]: How about `!this.currentInput`?

File: components/EditableTagList.vue [hunk 29]
```diff
@@ -118,14 +124,17 @@ export default class EditableTagList extends Vue {
...
     if (this.currentInput.length) {
       this.saveChip()
+      this.$nuxt.$nextTick(() => this.$refs.input.focus())
     } else {
       this.shouldShowInput = false
     }
   }
 
   saveChip() {
+    if(this.currentInput === '') return
```
Comment (williamchong) [inline components/EditableTagList.vue:137]: How about `!this.currentInput`?

File: pages/view/_iscnId.vue [hunk 30]
```diff
@@ -191,41 +395,108 @@ export default class ViewIscnIdPage extends Vue {
...
-          message: ErrorMessage.statusCode400,
-        })
-        return
+    } else {
+      const authorWalletAddress: any = []
+      authorWalletAddress.push({
+        type: 'cosmos',
+        address: stakeholders['@id'],
+      })
+      const authorUrl: any = []
```
Comment (williamchong) [inline pages/view/_iscnId.vue:458]: Should use plural form for arrays, i.e. `authorUrls`

File: pages/view/_iscnId.vue [hunk 31]
```diff
@@ -191,41 +395,108 @@ export default class ViewIscnIdPage extends Vue {
...
       }
-      const res = await this.fetchISCNByTx(param)
-      if (!res) {
-        this.$nuxt.error({
-          statusCode: 400,
-          message: ErrorMessage.statusCode400,
-        })
-        return
+    } else {
+      const authorWalletAddress: any = []
```
Comment (williamchong) [inline pages/view/_iscnId.vue:453]: Should use plural form for arrays, i.e. `authorWalletAddresses`

File: utils/cosmos/iscn/sign.ts [hunk 32]
```diff
@@ -25,31 +26,63 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+        }
+      })
+      if (authorUrls[i].length) {
+        authorUrls[i].forEach((a: any) => {
+          if (a.length) {
+            sameAs.push(a)
+          }
+        })
+      }
+      const isNonEmpty = url || authorName || identifier
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:72]: since `identifier` is an array, this is always `true` unless you use `identifier.length`

File: utils/cosmos/iscn/sign.ts [hunk 33]
```diff
@@ -25,31 +26,63 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+  const stakeholders: any = []
   if (authorNames.length) {
     for (let i = 0; i < authorNames.length; i += 1) {
-      const authorName = authorNames[i]
-      const authorUrl = authorUrls[i]
-      const authorId = authorWallets[i]
-      const isNonEmpty = authorUrl || authorName || authorId
+      const authorName: string = authorNames[i]
+      const sameAs: any = []
+      const identifier: any = []
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:44]: Prefer naming as `sameAsArray` `identifiers`, then when putting into object use the original name as key e.g. `identifier: identifiers`

File: utils/cosmos/iscn/sign.ts [hunk 34]
```diff
@@ -25,31 +26,63 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+          })
+        }
+      })
+      if (authorUrls[i].length) {
+        authorUrls[i].forEach((a: any) => {
+          if (a.length) {
+            sameAs.push(a)
+          }
+        })
+      }
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:71]: No need to check for `.length`, since if you are sure `authorUrls` is array (since you called `.length`), a forEach on empty array would not cause any issue.

Why need to check for `a.length`? What error case are we guarding?

**Review by williamchong — CHANGES_REQUESTED** [review]

File: utils/cosmos/iscn/sign.ts (hunk 34 above)
Comment (AuroraHuang22) [reply utils/cosmos/iscn/sign.ts:71]: 本來會一直 push 一個 empty string 上去 sameAs..... 結果剛剛重新 yarn dav 就沒問題了

File: pages/view/_iscnId.vue [hunk 35]
```diff
@@ -191,41 +395,111 @@ export default class ViewIscnIdPage extends Vue {
...
+        ? stakeholders.url.slice(16)
+        : ''
+      const authorWalletAddresses: any = []
+      stakeholders.identifier.forEach((a: any) => {
+        authorWalletAddresses.push({
+          address: a.value,
+          type: this.getKeyByValue(WALLET_TYPE_REPLACER, a.propertyID),
         })
-        return
+      })
```
Comment (williamchong) [inline pages/view/_iscnId.vue:443]: Since its onc-to-one array mapping, you can use `map()`
```
const authorWalletAddresses = stakeholders.identifier.map((a: any) => {
        return {
          address: a.value,
          type: this.getKeyByValue(WALLET_TYPE_REPLACER, a.propertyID),
        }
      })
```

File: pages/view/_iscnId.vue (hunk 27 above)
Comment (williamchong) [reply pages/view/_iscnId.vue:485]: seems we are still hardcoding `eth` here?

File: utils/cosmos/iscn/sign.ts [hunk 36]
```diff
@@ -25,31 +26,61 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+            propertyID: WALLET_TYPE_REPLACER[a.type],
+            value: `did:cosmos:${a.address.slice(6)}`,
+          })
+        } else {
+          identifiers.push({
+            '@type': 'PropertyValue',
+            propertyID: WALLET_TYPE_REPLACER[a.type],
+            value: `did:${a.type}:${a.address}`,
+          })
+        }
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:63]: We can also use `.map()` here

File: utils/cosmos/iscn/sign.ts [hunk 37]
```diff
@@ -25,31 +26,61 @@ export function formatISCNTxPayload(payload: ISCNRegisterPayload): ISCNSignPaylo
...
+      })
+      authorUrls[i].forEach((a: any) => {
+        if (a) {
+          sameAsArray.push(a)
+        }
+      })
```
Comment (williamchong) [inline utils/cosmos/iscn/sign.ts:69]: `We can use `.filter()` to filter out empty value, then again `.map()`

**Review by williamchong — CHANGES_REQUESTED** [review]

File: components/WalletFieldList.vue [hunk 38]
```diff
@@ -0,0 +1,185 @@
...
+          >
+            <IconEthereum />
+          </div>
+        </div>
+      </button>
+      <TextField
+        v-model="items.content"
+        :size="40"
+        :class="['my-[4px]', 'flex-grow']"
+        :placeholder="items.type === 'cosmos' ? 'comos....' : '0x.....'"
```
Comment (williamchong) [inline components/WalletFieldList.vue:84]:
```suggestion
        :placeholder="items.type === 'cosmos' ? 'cosmos....' : '0x.....'"
```

File: components/WalletFieldList.vue [hunk 39]
```diff
@@ -0,0 +1,185 @@
...
+            items-center
+            justify-center
+          "
+          @click="
+            items.isOpenOptions = !items.isOpenOptions
+            isSelecting = !isSelecting"
+        >
+          <div class="w-[20px] mx-[8px]">
+            <IconLike v-if="items.type === 'cosmos'" />
+            <IconEthereum v-if="items.type === 'eth'" />
```
Comment (williamchong) [inline components/WalletFieldList.vue:41]: Confusing that the icon for LIKE is used for cosmos, but we don't have an extra cosmos icon for cosmos

**Review by williamchong — CHANGES_REQUESTED** [review]
image
This should be a ETH address, not ERC20

image
Would be nice if the id after register is clickable

image
Should add address format validation check to prevent strange entry like this (mistype eth address into cosmos), or even better, smart identification of address type according to pattern

File: components/WalletFieldList.vue (hunk 39 above)
Comment (nwingt) [reply components/WalletFieldList.vue]: Asking @joshkiu to provide icon for cosmos type

**Comment by nwingt** [comment]
> image
> This should be a ETH address, not ERC20
> 
> image
> Would be nice if the id after register is clickable
> 
> image
> Should add address format validation check to prevent strange entry like this (mistype eth address into cosmos), or even better, smart identification of address type according to pattern

I think this kind of issue should be resolved in another task

**Review by williamchong — APPROVED** [review]
